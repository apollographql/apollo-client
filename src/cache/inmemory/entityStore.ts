import { dep, OptimisticDependencyFunction } from 'optimism';
import { invariant } from 'ts-invariant';
import { isReference, StoreValue } from '../../utilities/graphql/storeUtils';
import {
  DeepMerger,
  ReconcilerFunction,
} from '../../utilities/common/mergeDeep';
import { isEqual } from '../../utilities/common/isEqual';
import { NormalizedCache, NormalizedCacheObject, StoreObject } from './types';
import { getTypenameFromStoreObject } from './helpers';

const hasOwn = Object.prototype.hasOwnProperty;

type DependType = OptimisticDependencyFunction<string> | null;

function makeDepKey(dataId: string, fieldName?: string) {
  const parts = [dataId];
  if (typeof fieldName === "string") {
    parts.push(fieldName);
  }
  return JSON.stringify(parts);
}

function depend(store: EntityStore, dataId: string, fieldName?: string) {
  if (store.depend) {
    store.depend(makeDepKey(dataId, fieldName));
  }
}

function dirty(store: EntityStore, dataId: string, fieldName?: string) {
  if (store.depend) {
    store.depend.dirty(makeDepKey(dataId));
    if (typeof fieldName === "string") {
      store.depend.dirty(makeDepKey(dataId, fieldName));
    }
  }
}

export abstract class EntityStore implements NormalizedCache {
  protected data: NormalizedCacheObject = Object.create(null);

  // It seems like this property ought to be protected rather than public,
  // but TypeScript doesn't realize it's inherited from a shared base
  // class by both Root and Layer classes, so Layer methods are forbidden
  // from accessing the .depend property of an arbitrary EntityStore
  // instance, because it might be a Root instance (and vice-versa).
  public readonly depend: DependType = null;

  public abstract addLayer(
    layerId: string,
    replay: (layer: EntityStore) => any,
  ): EntityStore;

  public abstract removeLayer(layerId: string): EntityStore;

  // Although the EntityStore class is abstract, it contains concrete
  // implementations of the various NormalizedCache interface methods that
  // are inherited by the Root and Layer subclasses.

  public toObject(): NormalizedCacheObject {
    return { ...this.data };
  }

  public has(dataId: string): boolean {
    return this.get(dataId) !== void 0;
  }

  public get(dataId: string): StoreObject {
    depend(this, dataId);
    return this.data[dataId];
  }

  public getFieldValue(dataId: string, fieldName: string): StoreValue {
    depend(this, dataId, fieldName);
    const storeObject = this.data[dataId];
    return storeObject && storeObject[fieldName];
  }

  public merge(dataId: string, incoming: StoreObject): void {
    const existing = this.get(dataId);
    const merged = new DeepMerger(storeObjectReconciler)
      .merge(existing, incoming, this);
    if (merged !== existing) {
      this.data[dataId] = merged;
      delete this.refs[dataId];
      if (this.depend) {
        // First, invalidate any dependents that called get rather than
        // getFieldValue.
        dirty(this, dataId);
        // Now invalidate dependents who called getFieldValue for any
        // fields that are changing as a result of this merge.
        Object.keys(incoming).forEach(fieldName => {
          if (!existing || incoming[fieldName] !== existing[fieldName]) {
            dirty(this, dataId, fieldName);
          }
        });
      }
    }
  }

  // TODO Allow deleting fields of this.data[dataId] according to their
  // original field.name.value.
  public delete(dataId: string): void {
    const storeObject = this.data[dataId];

    delete this.data[dataId];
    delete this.refs[dataId];
    // Note that we do not delete the this.rootIds[dataId] retainment
    // count for this ID, since an object with the same ID could appear in
    // the store again, and should not have to be retained again.
    // delete this.rootIds[dataId];

    if (this.depend && storeObject) {
      dirty(this, dataId);
      Object.keys(storeObject).forEach(fieldName => {
        dirty(this, dataId, fieldName);
      });
    }
  }

  public clear(): void {
    this.replace(null);
  }

  public replace(newData: NormalizedCacheObject | null): void {
    Object.keys(this.data).forEach(dataId => {
      if (!(newData && hasOwn.call(newData, dataId))) {
        this.delete(dataId);
      }
    });
    if (newData) {
      Object.keys(newData).forEach(dataId => {
        this.merge(dataId, newData[dataId]);
      });
    }
  }

  // Maps root entity IDs to the number of times they have been retained, minus
  // the number of times they have been released. Retained entities keep other
  // entities they reference (even indirectly) from being garbage collected.
  private rootIds: {
    [rootId: string]: number;
  } = Object.create(null);

  public retain(rootId: string): number {
    return this.rootIds[rootId] = (this.rootIds[rootId] || 0) + 1;
  }

  public release(rootId: string): number {
    if (this.rootIds[rootId] > 0) {
      const count = --this.rootIds[rootId];
      if (!count) delete this.rootIds[rootId];
      return count;
    }
    return 0;
  }

  // This method will be overridden in the Layer class to merge root IDs for all
  // layers (including the root).
  public getRootIdSet() {
    return new Set(Object.keys(this.rootIds));
  }

  // The goal of garbage collection is to remove IDs from the Root layer of the
  // store that are no longer reachable starting from any IDs that have been
  // explicitly retained (see retain and release, above). Returns an array of
  // dataId strings that were removed from the store.
  public gc() {
    const ids = this.getRootIdSet();
    const snapshot = this.toObject();
    ids.forEach(id => {
      if (hasOwn.call(snapshot, id)) {
        // Because we are iterating over an ECMAScript Set, the IDs we add here
        // will be visited in later iterations of the forEach loop only if they
        // were not previously contained by the Set.
        Object.keys(this.findChildRefIds(id)).forEach(ids.add, ids);
        // By removing IDs from the snapshot object here, we protect them from
        // getting removed from the root store layer below.
        delete snapshot[id];
      }
    });
    const idsToRemove = Object.keys(snapshot);
    if (idsToRemove.length) {
      let root: EntityStore = this;
      while (root instanceof Layer) root = root.parent;
      idsToRemove.forEach(root.delete, root);
    }
    return idsToRemove;
  }

  // Lazily tracks { __ref: <dataId> } strings contained by this.data[dataId].
  private refs: {
    [dataId: string]: Record<string, true>;
  } = Object.create(null);

  public findChildRefIds(dataId: string): Record<string, true> {
    if (!hasOwn.call(this.refs, dataId)) {
      const found = this.refs[dataId] = Object.create(null);
      const workSet = new Set([this.data[dataId]]);
      // Within the store, only arrays and objects can contain child entity
      // references, so we can prune the traversal using this predicate:
      const canTraverse = (obj: any) => obj !== null && typeof obj === 'object';
      workSet.forEach(obj => {
        if (isReference(obj)) {
          found[obj.__ref] = true;
        } else if (canTraverse(obj)) {
          Object.values(obj)
            // No need to add primitive values to the workSet, since they cannot
            // contain reference objects.
            .filter(canTraverse)
            .forEach(workSet.add, workSet);
        }
      });
    }
    return this.refs[dataId];
  }
}

export namespace EntityStore {
  // Refer to this class as EntityStore.Root outside this namespace.
  export class Root extends EntityStore {
    // Although each Root instance gets its own unique this.depend
    // function, any Layer instances created by calling addLayer need to
    // share a single distinct dependency function. Since this shared
    // function must outlast the Layer instances themselves, it needs to
    // be created and owned by the Root instance.
    private sharedLayerDepend: DependType = null;

    constructor({
      resultCaching = true,
      seed,
    }: {
      resultCaching?: boolean;
      seed?: NormalizedCacheObject;
    }) {
      super();
      if (resultCaching) {
        // Regard this.depend as publicly readonly but privately mutable.
        (this as any).depend = dep<string>();
        this.sharedLayerDepend = dep<string>();
      }
      if (seed) this.replace(seed);
    }

    public addLayer(
      layerId: string,
      replay: (layer: EntityStore) => any,
    ): EntityStore {
      // The replay function will be called in the Layer constructor.
      return new Layer(layerId, this, replay, this.sharedLayerDepend);
    }

    public removeLayer(layerId: string): Root {
      // Never remove the root layer.
      return this;
    }
  }
}

// Not exported, since all Layer instances are created by the addLayer method
// of the EntityStore.Root class.
class Layer extends EntityStore {
  constructor(
    public readonly id: string,
    public readonly parent: Layer | EntityStore.Root,
    public readonly replay: (layer: EntityStore) => any,
    public readonly depend: DependType,
  ) {
    super();
    replay(this);
  }

  public addLayer(
    layerId: string,
    replay: (layer: EntityStore) => any,
  ): EntityStore {
    return new Layer(layerId, this, replay, this.depend);
  }

  public removeLayer(layerId: string): EntityStore {
    // Remove all instances of the given id, not just the first one.
    const parent = this.parent.removeLayer(layerId);

    if (layerId === this.id) {
      // Dirty every ID we're removing.
      // TODO Some of these IDs could escape dirtying if value unchanged.
      if (this.depend) {
        Object.keys(this.data).forEach(dataId => this.delete(dataId));
      }
      return parent;
    }

    // No changes are necessary if the parent chain remains identical.
    if (parent === this.parent) return this;

    // Recreate this layer on top of the new parent.
    return parent.addLayer(this.id, this.replay);
  }

  public toObject(): NormalizedCacheObject {
    return {
      ...this.parent.toObject(),
      ...this.data,
    };
  }

  public get(dataId: string): StoreObject {
    if (hasOwn.call(this.data, dataId)) {
      return super.get(dataId);
    }

    // If this layer has a this.depend function and it's not the one
    // this.parent is using, we need to depend on the given dataId using
    // this.depend before delegating to the parent. This check saves us
    // from calling this.depend for every optimistic layer we examine, but
    // ensures we call this.depend in the last optimistic layer before we
    // reach the root layer.
    if (this.depend && this.depend !== this.parent.depend) {
      depend(this, dataId);
    }

    return this.parent.get(dataId);
  }

  public getFieldValue(dataId: string, fieldName: string): StoreValue {
    if (hasOwn.call(this.data, dataId)) {
      const storeObject = this.data[dataId];
      if (storeObject && hasOwn.call(storeObject, fieldName)) {
        return super.getFieldValue(dataId, fieldName);
      }
    }

    if (this.depend && this.depend !== this.parent.depend) {
      depend(this, dataId, fieldName);
    }

    return this.parent.getFieldValue(dataId, fieldName);
  }

  public delete(dataId: string): void {
    super.delete(dataId);
    // In case this.parent (or one of its ancestors) has an entry for this ID,
    // we need to shadow it with an undefined value, or it might be inherited
    // by the Layer#get method.
    this.data[dataId] = void 0;
  }

  // Return a Set<string> of all the ID strings that have been retained by this
  // Layer *and* any layers/roots beneath it.
  public getRootIdSet(): Set<string> {
    const ids = this.parent.getRootIdSet();
    super.getRootIdSet().forEach(ids.add, ids);
    return ids;
  }

  public findChildRefIds(dataId: string): Record<string, true> {
    const fromParent = this.parent.findChildRefIds(dataId);
    return hasOwn.call(this.data, dataId) ? {
      ...fromParent,
      ...super.findChildRefIds(dataId),
    } : fromParent;
  }
}

const storeObjectReconciler: ReconcilerFunction<[EntityStore]> = function (
  existingObject,
  incomingObject,
  property,
  // This parameter comes from the additional argument we pass to the
  // merge method in context.mergeStoreObjects (see writeQueryToStore).
  store,
) {
  // In the future, reconciliation logic may depend on the type of the parent
  // StoreObject, not just the values of the given property.
  const existing = existingObject[property];
  const incoming = incomingObject[property];

  if (
    existing !== incoming &&
    // The DeepMerger class has various helpful utilities that we might as
    // well reuse here.
    this.isObject(existing) &&
    this.isObject(incoming)
  ) {
    const eType = getTypenameFromStoreObject(store, existing);
    const iType = getTypenameFromStoreObject(store, incoming);
    // If both objects have a typename and the typename is different, let the
    // incoming object win. The typename can change when a different subtype
    // of a union or interface is written to the store.
    if (
      typeof eType === 'string' &&
      typeof iType === 'string' &&
      eType !== iType
    ) {
      return incoming;
    }

    invariant(
      !isReference(existing) || isReference(incoming),
      `Store error: the application attempted to write an object with no provided id but the store already contains an id of ${existing.__ref} for this object.`,
    );

    // It's worth checking deep equality here (even though blindly
    // returning incoming would be logically correct) because preserving
    // the referential identity of existing data can prevent needless
    // rereading and rerendering.
    if (isEqual(existing, incoming)) {
      return existing;
    }
  }

  // In all other cases, incoming replaces existing without any effort to
  // merge them deeply, since custom merge functions have already been
  // applied to the incoming data by walkWithMergeOverrides.
  return incoming;
}

export function supportsResultCaching(store: any): store is EntityStore {
  // When result caching is disabled, store.depend will be null.
  return !!(store instanceof EntityStore && store.depend);
}

export function defaultNormalizedCacheFactory(
  seed?: NormalizedCacheObject,
): NormalizedCache {
  return new EntityStore.Root({ resultCaching: true, seed });
}
