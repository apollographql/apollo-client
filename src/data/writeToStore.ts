import {
  isString,
  isNumber,
  isBoolean,
  isNull,
  isArray,
  isUndefined,
  assign,
} from 'lodash';

import {
  parseFragment,
  parseQuery,
} from '../parser';

import {
  storeKeyNameFromField,
  resultKeyNameFromField,
} from './storeUtils';

import {
  OperationDefinition,
  SelectionSet,
  Selection,
  FragmentDefinition,
  Field,
  InlineFragment,
} from 'graphql';

import {
  NormalizedCache,
  StoreObject,
} from './store';

import {
  IdGetter,
} from './extensions';

// import {
//   printAST,
// } from './debug';

/**
 * Convert a nested GraphQL result into a normalized store, where each object from the schema
 * appears exactly once.
 * @param  {Object} result Arbitrary nested JSON, returned from the GraphQL server
 * @param  {String} [fragment] The GraphQL fragment used to fetch the data in result
 * @param  {SelectionSet} [selectionSet] The parsed selection set for the subtree of the query this
 *                                       result represents
 * @param  {Object} [store] The store to merge into
 * @return {Object} The resulting store
 */
export function writeFragmentToStore({
  result,
  fragment,
  store = {} as NormalizedCache,
  variables,
  dataIdFromObject = null,
}: {
  result: Object,
  fragment: string,
  store?: NormalizedCache,
  variables?: Object,
  dataIdFromObject?: IdGetter,
}): NormalizedCache {
  // Argument validation
  if (!fragment) {
    throw new Error('Must pass fragment.');
  }

  const parsedFragment: FragmentDefinition = parseFragment(fragment);
  const selectionSet: SelectionSet = parsedFragment.selectionSet;

  if (!result['id']) {
    throw new Error('Result must have id when writing fragment to store.');
  }

  return writeSelectionSetToStore({
    dataId: result['id'],
    result,
    selectionSet,
    store,
    variables,
    dataIdFromObject,
  });
}

export function writeQueryToStore({
  result,
  query,
  store = {} as NormalizedCache,
  variables,
  dataIdFromObject = null,
}: {
  result: Object,
  query: string,
  store?: NormalizedCache,
  variables?: Object,
  dataIdFromObject?: IdGetter,
}): NormalizedCache {
  const queryDefinition: OperationDefinition = parseQuery(query);

  return writeSelectionSetToStore({
    dataId: 'ROOT_QUERY',
    result,
    selectionSet: queryDefinition.selectionSet,
    store,
    variables,
    dataIdFromObject,
  });
}

export function writeSelectionSetToStore({
  result,
  dataId,
  selectionSet,
  store = {} as NormalizedCache,
  variables,
  dataIdFromObject,
}: {
  dataId: string,
  result: any,
  selectionSet: SelectionSet,
  store?: NormalizedCache,
  variables: Object,
  dataIdFromObject: IdGetter,
}): NormalizedCache {
  selectionSet.selections.forEach((selection) => {
    if (isField(selection)) {
      const resultFieldKey: string = resultKeyNameFromField(selection);
      const value: any = result[resultFieldKey];

      if (isUndefined(value)) {
        throw new Error(`Can't find field ${resultFieldKey} on result object ${dataId}.`);
      }

      writeFieldToStore({
        dataId,
        value,
        variables,
        store,
        field: selection,
        dataIdFromObject,
      });
    } else if (isInlineFragment(selection)) {
      // XXX what to do if this tries to write the same fields? Also, type conditions...
      writeSelectionSetToStore({
        result,
        selectionSet: selection.selectionSet,
        store,
        variables,
        dataId,
        dataIdFromObject,
      });
    } else {
      throw new Error('Non-inline fragments not supported.');
    }
  });

  return store;
}

function writeFieldToStore({
  field,
  value,
  variables,
  store,
  dataId,
  dataIdFromObject,
}: {
  field: Field,
  value: any,
  variables: {},
  store: NormalizedCache,
  dataId: string,
  dataIdFromObject: IdGetter,
}) {
  let storeValue;

  const storeFieldName: string = storeKeyNameFromField(field, variables);

  // If it's a scalar, just store it in the store
  if (isString(value) || isNumber(value) || isBoolean(value) || isNull(value)) {
    storeValue = value;
  } else if (isArray(value)) {

    // GraphQL lists should be of the same type.
    // If it's an array of scalar values, don't normalize.
    if (isNull(field.selectionSet)) {
      storeValue = value;
    } else {

      const thisIdList: Array<string> = [];

      value.forEach((item, index) => {
        if (isNull(item)) {
          thisIdList.push(null);
        } else {
          let itemDataId = `${dataId}.${storeFieldName}.${index}`;

          if (dataIdFromObject) {
            const semanticId = dataIdFromObject(item);

            if (semanticId) {
              itemDataId = semanticId;
            }
          }

          thisIdList.push(itemDataId);

          writeSelectionSetToStore({
            dataId: itemDataId,
            result: item,
            store,
            selectionSet: field.selectionSet,
            variables,
            dataIdFromObject,
          });
        }
      });

      storeValue = thisIdList;
    }
  } else {
    // It's an object
    let valueDataId = `${dataId}.${storeFieldName}`;

    if (dataIdFromObject) {
      const semanticId = dataIdFromObject(value);

      if (semanticId) {
        valueDataId = semanticId;
      }
    }

    writeSelectionSetToStore({
      dataId: valueDataId,
      result: value,
      store,
      selectionSet: field.selectionSet,
      variables,
      dataIdFromObject,
    });

    storeValue = valueDataId;
  }

  const newStoreObj = assign({}, store[dataId], {
    [storeFieldName]: storeValue,
  }) as StoreObject;

  store[dataId] = newStoreObj;
}

function isField(selection: Selection): selection is Field {
  return selection.kind === 'Field';
}

function isInlineFragment(selection: Selection): selection is InlineFragment {
  return selection.kind === 'InlineFragment';
}
