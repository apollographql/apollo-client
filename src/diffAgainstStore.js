import {
  isArray,
  has,
} from 'lodash';

import {
  parseFragmentIfString,
  parseQueryIfString,
} from './parser';

import {
  cacheFieldNameFromSelection,
  resultFieldNameFromSelection,
} from './cacheUtils';

export function diffQueryAgainstStore({ store, query }) {
  const queryDef = parseQueryIfString(query);

  return diffSelectionSetAgainstStore({
    store,
    rootId: 'ROOT_QUERY',
    selectionSet: queryDef.selectionSet,
  });
}

export function diffFragmentAgainstStore({ store, fragment, rootId }) {
  const fragmentDef = parseFragmentIfString(fragment);

  return diffSelectionSetAgainstStore({
    store,
    rootId,
    selectionSet: fragmentDef.selectionSet,
  });
}

export function diffSelectionSetAgainstStore({
  selectionSet,
  store,
  rootId,
  throwOnMissingField = false,
}) {
  if (selectionSet.kind !== 'SelectionSet') {
    throw new Error('Must be a selection set.');
  }

  const result = {};
  const missingFields = [];

  const cacheObj = store[rootId];

  selectionSet.selections.forEach((selection) => {
    const cacheFieldName = cacheFieldNameFromSelection(selection);
    const resultFieldName = resultFieldNameFromSelection(selection);

    if (! has(cacheObj, cacheFieldName)) {
      if (throwOnMissingField) {
        throw new Error(`Can't find field ${cacheFieldName} on object ${cacheObj}.`);
      }

      missingFields.push({
        id: rootId,
        field: cacheFieldName,
      });
      return;
    }

    if (! selection.selectionSet) {
      result[resultFieldName] = cacheObj[cacheFieldName];
      return;
    }

    if (isArray(cacheObj[cacheFieldName])) {
      result[resultFieldName] = cacheObj[cacheFieldName].map((id) => {
        const itemDiffResult = diffSelectionSetAgainstStore({
          store,
          throwOnMissingField,
          rootId: id,
          selectionSet: selection.selectionSet,
        });

        itemDiffResult.missingFields.forEach(field => missingFields.push(field));
        return itemDiffResult.result;
      });
      return;
    }

    const subObjDiffResult = diffSelectionSetAgainstStore({
      store,
      throwOnMissingField,
      rootId: cacheObj[cacheFieldName],
      selectionSet: selection.selectionSet,
    });

    // This is a nested query
    subObjDiffResult.missingFields.forEach(field => missingFields.push(field));
    result[resultFieldName] = subObjDiffResult.result;
  });

  return {
    result,
    missingFields,
  };
}
