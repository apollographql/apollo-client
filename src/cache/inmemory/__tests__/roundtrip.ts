import { DocumentNode } from 'graphql';
import gql from 'graphql-tag';

import { withError } from './diffAgainstStore';
import { EntityStore } from '../entityStore';
import { StoreReader } from '../readFromStore';
import { StoreWriter } from '../writeToStore';
import { Policies } from '../policies';

function assertDeeplyFrozen(value: any, stack: any[] = []) {
  if (value !== null && typeof value === 'object' && stack.indexOf(value) < 0) {
    expect(Object.isExtensible(value)).toBe(false);
    expect(Object.isFrozen(value)).toBe(true);
    stack.push(value);
    Object.keys(value).forEach(key => {
      assertDeeplyFrozen(value[key], stack);
    });
    expect(stack.pop()).toBe(value);
  }
}

function storeRoundtrip(query: DocumentNode, result: any, variables = {}) {
  const policies = new Policies({
    possibleTypes: {
      Character: ["Jedi", "Droid"],
    },
  });

  const reader = new StoreReader({ policies });
  const writer = new StoreWriter({ policies });

  const store = writer.writeQueryToStore({
    result,
    query,
    variables,
  });

  const readOptions = {
    store,
    query,
    variables,
  };

  const reconstructedResult = reader.readQueryFromStore<any>(readOptions);
  expect(reconstructedResult).toEqual(result);

  // Make sure the result is identical if we haven't written anything new
  // to the store. https://github.com/apollographql/apollo-client/pull/3394
  expect(store).toBeInstanceOf(EntityStore);
  expect(reader.readQueryFromStore(readOptions)).toBe(reconstructedResult);

  const immutableResult = reader.readQueryFromStore(readOptions);
  expect(immutableResult).toEqual(reconstructedResult);
  expect(reader.readQueryFromStore(readOptions)).toBe(immutableResult);
  if (process.env.NODE_ENV !== 'production') {
    try {
      // Note: this illegal assignment will only throw in strict mode, but that's
      // safe to assume because this test file is a module.
      (immutableResult as any).illegal = 'this should not work';
      throw new Error('unreached');
    } catch (e) {
      expect(e.message).not.toMatch(/unreached/);
      expect(e).toBeInstanceOf(TypeError);
    }
    assertDeeplyFrozen(immutableResult);
  }

  // Now make sure subtrees of the result are identical even after we write
  // an additional bogus field to the store.
  writer.writeQueryToStore({
    store,
    result: { oyez: 1234 },
    query: gql`
      {
        oyez
      }
    `,
  });

  const deletedRootResult = reader.readQueryFromStore<any>(readOptions);
  expect(deletedRootResult).toEqual(result);

  if (deletedRootResult === reconstructedResult) {
    // We don't expect the new result to be identical to the previous result,
    // but there are some rare cases where that can happen, and it's a good
    // thing, because it means the caching system is working slightly better
    // than expected... and we don't need to continue with the rest of the
    // comparison logic below.
    return;
  }

  Object.keys(result).forEach(key => {
    expect(deletedRootResult[key]).toBe(reconstructedResult[key]);
  });
}

describe('roundtrip', () => {
  it('real graphql result', () => {
    storeRoundtrip(
      gql`
        {
          people_one(id: "1") {
            name
          }
        }
      `,
      {
        people_one: {
          name: 'Luke Skywalker',
        },
      },
    );
  });

  it('multidimensional array (#776)', () => {
    storeRoundtrip(
      gql`
        {
          rows {
            value
          }
        }
      `,
      {
        rows: [[{ value: 1 }, { value: 2 }], [{ value: 3 }, { value: 4 }]],
      },
    );
  });

  it('array with null values (#1551)', () => {
    storeRoundtrip(
      gql`
        {
          list {
            value
          }
        }
      `,
      {
        list: [null, { value: 1 }],
      },
    );
  });

  it('enum arguments', () => {
    storeRoundtrip(
      gql`
        {
          hero(episode: JEDI) {
            name
          }
        }
      `,
      {
        hero: {
          name: 'Luke Skywalker',
        },
      },
    );
  });

  it('with an alias', () => {
    storeRoundtrip(
      gql`
        {
          luke: people_one(id: "1") {
            name
          }
          vader: people_one(id: "4") {
            name
          }
        }
      `,
      {
        luke: {
          name: 'Luke Skywalker',
        },
        vader: {
          name: 'Darth Vader',
        },
      },
    );
  });

  it('with variables', () => {
    storeRoundtrip(
      gql`
        {
          luke: people_one(id: $lukeId) {
            name
          }
          vader: people_one(id: $vaderId) {
            name
          }
        }
      `,
      {
        luke: {
          name: 'Luke Skywalker',
        },
        vader: {
          name: 'Darth Vader',
        },
      },
      {
        lukeId: '1',
        vaderId: '4',
      },
    );
  });

  it('with GraphQLJSON scalar type', () => {
    const updateClub = {
      uid: '1d7f836018fc11e68d809dfee940f657',
      name: 'Eple',
      settings: {
        name: 'eple',
        currency: 'AFN',
        calendarStretch: 2,
        defaultPreAllocationPeriod: 1,
        confirmationEmailCopy: null,
        emailDomains: null,
      },
    } as any;

    storeRoundtrip(
      gql`
        {
          updateClub {
            uid
            name
            settings
          }
        }
      `,
      {
        updateClub,
      },
    );

    // Reading immutable results from the store does not mean the original
    // data should get frozen.
    expect(Object.isExtensible(updateClub)).toBe(true);
    expect(Object.isFrozen(updateClub)).toBe(false);
  });

  describe('directives', () => {
    it('should be able to query with skip directive true', () => {
      storeRoundtrip(
        gql`
          query {
            fortuneCookie @skip(if: true)
          }
        `,
        {},
      );
    });

    it('should be able to query with skip directive false', () => {
      storeRoundtrip(
        gql`
          query {
            fortuneCookie @skip(if: false)
          }
        `,
        { fortuneCookie: 'live long and prosper' },
      );
    });
  });

  describe('fragments', () => {
    it('should work on null fields', () => {
      storeRoundtrip(
        gql`
          query {
            field {
              ... on Obj {
                stuff
              }
            }
          }
        `,
        {
          field: null,
        },
      );
    });

    it('should work on basic inline fragments', () => {
      storeRoundtrip(
        gql`
          query {
            field {
              __typename
              ... on Obj {
                stuff
              }
            }
          }
        `,
        {
          field: {
            __typename: 'Obj',
            stuff: 'Result',
          },
        },
      );
    });

    // XXX this test is weird because it assumes the server returned an incorrect result
    // However, the user may have written this result with client.writeQuery.
    it('should throw an error on two of the same inline fragment types', () => {
      expect(() => {
        storeRoundtrip(
          gql`
            query {
              all_people {
                __typename
                name
                ... on Jedi {
                  side
                }
                ... on Jedi {
                  rank
                }
              }
            }
          `,
          {
            all_people: [
              {
                __typename: 'Jedi',
                name: 'Luke Skywalker',
                side: 'bright',
              },
            ],
          },
        );
      }).toThrowError(/Missing field 'rank' /);
    });

    it('should resolve fields it can on interface with non matching inline fragments', () => {
      return withError(() => {
        storeRoundtrip(
          gql`
            query {
              dark_forces {
                __typename
                name
                ... on Droid {
                  model
                }
              }
            }
          `,
          {
            dark_forces: [
              {
                __typename: 'Droid',
                name: '8t88',
                model: '88',
              },
              {
                __typename: 'Darth',
                name: 'Anakin Skywalker',
              },
            ],
          },
        );
      });
    });

    it('should resolve on union types with spread fragments', () => {
      return withError(() => {
        storeRoundtrip(
          gql`
            fragment jediFragment on Jedi {
              side
            }

            fragment droidFragment on Droid {
              model
            }

            query {
              all_people {
                __typename
                name
                ...jediFragment
                ...droidFragment
              }
            }
          `,
          {
            all_people: [
              {
                __typename: 'Jedi',
                name: 'Luke Skywalker',
                side: 'bright',
              },
              {
                __typename: 'Droid',
                name: 'R2D2',
                model: 'astromech',
              },
            ],
          },
        );
      });
    });

    it('should work with a fragment on the actual interface or union', () => {
      return withError(() => {
        storeRoundtrip(
          gql`
            fragment jediFragment on Character {
              side
            }

            fragment droidFragment on Droid {
              model
            }

            query {
              all_people {
                name
                __typename
                ...jediFragment
                ...droidFragment
              }
            }
          `,
          {
            all_people: [
              {
                __typename: 'Jedi',
                name: 'Luke Skywalker',
                side: 'bright',
              },
              {
                __typename: 'Droid',
                name: 'R2D2',
                model: 'astromech',
                side: 'bright',
              },
            ],
          },
        );
      });
    });

    it('should throw on error on two of the same spread fragment types', () => {
      expect(() => {
        storeRoundtrip(
          gql`
            fragment jediSide on Jedi {
              side
            }

            fragment jediRank on Jedi {
              rank
            }

            query {
              all_people {
                __typename
                name
                ...jediSide
                ...jediRank
              }
            }
          `,
          {
            all_people: [
              {
                __typename: 'Jedi',
                name: 'Luke Skywalker',
                side: 'bright',
              },
            ],
          },
        );
      }).toThrowError(/Missing field 'rank' /);
    });

    it('should resolve on @include and @skip with inline fragments', () => {
      storeRoundtrip(
        gql`
          query {
            person {
              name
              __typename
              ... on Jedi @include(if: true) {
                side
              }
              ... on Droid @skip(if: true) {
                model
              }
            }
          }
        `,
        {
          person: {
            __typename: 'Jedi',
            name: 'Luke Skywalker',
            side: 'bright',
          },
        },
      );
    });

    it('should resolve on @include and @skip with spread fragments', () => {
      storeRoundtrip(
        gql`
          fragment jediFragment on Jedi {
            side
          }

          fragment droidFragment on Droid {
            model
          }

          query {
            person {
              name
              __typename
              ...jediFragment @include(if: true)
              ...droidFragment @skip(if: true)
            }
          }
        `,
        {
          person: {
            __typename: 'Jedi',
            name: 'Luke Skywalker',
            side: 'bright',
          },
        },
      );
    });
  });
});
