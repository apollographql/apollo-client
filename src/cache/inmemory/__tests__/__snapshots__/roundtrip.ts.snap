// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`roundtrip fragments should throw an error on two of the same inline fragment types 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Missing field 'rank' while writing result {
  \\"__typename\\": \\"Jedi\\",
  \\"name\\": \\"Luke Skywalker\\",
  \\"side\\": \\"bright\\"
}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`roundtrip fragments should throw on error on two of the same spread fragment types 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Missing field 'rank' while writing result {
  \\"__typename\\": \\"Jedi\\",
  \\"name\\": \\"Luke Skywalker\\",
  \\"side\\": \\"bright\\"
}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
